### ISSUE TITLE
orders (full): bad query performance for sales_shipment_track data

### ISSUE THREAD

**Issue by heldchen** [boxalino/data-integration-magento2#2]
on large enough instances the query performance when reading the shipment data is very bad and forces mysql to use filebased filtering & sorting. the bad query is using nested `GROUP BY` without filter clauses:

```
SELECT s_o.entity_id AS di_id, s_o.*, s_o_t.*, s_o_p.*, s_s_t.* FROM ( SELECT s_o.* FROM sales_order AS s_o WHERE (s_o.store_id IN ('1', '2') OR s_o.store_id = 0) ORDER BY s_o.entity_id DESC
 LIMIT 10 OFFSET 10 ) AS s_o

 LEFT JOIN ( SELECT sales_order_tax.tax_id AS s_o_t_tax_id, sales_order_tax.order_id AS s_o_t_order_id, sales_order_tax.code AS s_o_t_code, sales_order_tax.title AS s_o_t_title, sales_order_tax.percent AS s_o_t_percent, sales_order_tax.amount AS s_o_t_amount, sales_order_tax.priority AS s_o_t_priority, sales_order_tax.position AS s_o_t_position, sales_order_tax.base_amount AS s_o_t_base_amount, sales_order_tax.process AS s_o_t_process, sales_order_tax.base_real_amount AS s_o_t_base_real_amount FROM sales_order_tax GROUP BY order_id ) AS s_o_t ON s_o_t.s_o_t_order_id = s_o.entity_id

 LEFT JOIN ( SELECT sales_order_payment.entity_id AS s_o_p_entity_id, sales_order_payment.parent_id AS s_o_p_parent_id, sales_order_payment.base_shipping_captured AS s_o_p_base_shipping_captured, sales_order_payment.shipping_captured AS s_o_p_shipping_captured, sales_order_payment.amount_refunded AS s_o_p_amount_refunded, sales_order_payment.base_amount_paid AS s_o_p_base_amount_paid, sales_order_payment.amount_canceled AS s_o_p_amount_canceled, sales_order_payment.base_amount_authorized AS s_o_p_base_amount_authorized, sales_order_payment.base_amount_paid_online AS s_o_p_base_amount_paid_online, sales_order_payment.base_amount_refunded_online AS s_o_p_base_amount_refunded_online, sales_order_payment.base_shipping_amount AS s_o_p_base_shipping_amount, sales_order_payment.shipping_amount AS s_o_p_shipping_amount, sales_order_payment.amount_paid AS s_o_p_amount_paid, sales_order_payment.amount_authorized AS s_o_p_amount_authorized, sales_order_payment.base_amount_ordered AS s_o_p_base_amount_ordered, sales_order_payment.base_shipping_refunded AS s_o_p_base_shipping_refunded, sales_order_payment.shipping_refunded AS s_o_p_shipping_refunded, sales_order_payment.base_amount_refunded AS s_o_p_base_amount_refunded, sales_order_payment.amount_ordered AS s_o_p_amount_ordered, sales_order_payment.base_amount_canceled AS s_o_p_base_amount_canceled, sales_order_payment.quote_payment_id AS s_o_p_quote_payment_id, sales_order_payment.additional_data AS s_o_p_additional_data, sales_order_payment.cc_exp_month AS s_o_p_cc_exp_month, sales_order_payment.cc_ss_start_year AS s_o_p_cc_ss_start_year, sales_order_payment.echeck_bank_name AS s_o_p_echeck_bank_name, sales_order_payment.method AS s_o_p_method, sales_order_payment.cc_debug_request_body AS s_o_p_cc_debug_request_body, sales_order_payment.cc_secure_verify AS s_o_p_cc_secure_verify, sales_order_payment.protection_eligibility AS s_o_p_protection_eligibility, sales_order_payment.cc_approval AS s_o_p_cc_approval, sales_order_payment.cc_last_4 AS s_o_p_cc_last_4, sales_order_payment.cc_status_description AS s_o_p_cc_status_description, sales_order_payment.echeck_type AS s_o_p_echeck_type, sales_order_payment.cc_debug_response_serialized AS s_o_p_cc_debug_response_serialized, sales_order_payment.cc_ss_start_month AS s_o_p_cc_ss_start_month, sales_order_payment.echeck_account_type AS s_o_p_echeck_account_type, sales_order_payment.last_trans_id AS s_o_p_last_trans_id, sales_order_payment.cc_cid_status AS s_o_p_cc_cid_status, sales_order_payment.cc_owner AS s_o_p_cc_owner, sales_order_payment.cc_type AS s_o_p_cc_type, sales_order_payment.po_number AS s_o_p_po_number, sales_order_payment.cc_exp_year AS s_o_p_cc_exp_year, sales_order_payment.cc_status AS s_o_p_cc_status, sales_order_payment.echeck_routing_number AS s_o_p_echeck_routing_number, sales_order_payment.account_status AS s_o_p_account_status, sales_order_payment.anet_trans_method AS s_o_p_anet_trans_method, sales_order_payment.cc_debug_response_body AS s_o_p_cc_debug_response_body, sales_order_payment.cc_ss_issue AS s_o_p_cc_ss_issue, sales_order_payment.echeck_account_name AS s_o_p_echeck_account_name, sales_order_payment.cc_avs_status AS s_o_p_cc_avs_status, sales_order_payment.cc_number_enc AS s_o_p_cc_number_enc, sales_order_payment.cc_trans_id AS s_o_p_cc_trans_id, sales_order_payment.address_status AS s_o_p_address_status, sales_order_payment.additional_information AS s_o_p_additional_information FROM sales_order_payment GROUP BY parent_id ) AS s_o_p ON s_o_p.s_o_p_parent_id = s_o.entity_id

 LEFT JOIN ( SELECT sales_shipment_track.entity_id AS s_s_t_entity_id, sales_shipment_track.parent_id AS s_s_t_parent_id, sales_shipment_track.weight AS s_s_t_weight, sales_shipment_track.qty AS s_s_t_qty, sales_shipment_track.order_id AS s_s_t_order_id, sales_shipment_track.track_number AS s_s_t_track_number, sales_shipment_track.description AS s_s_t_description, sales_shipment_track.title AS s_s_t_title, sales_shipment_track.carrier_code AS s_s_t_carrier_code, sales_shipment_track.created_at AS s_s_t_created_at, sales_shipment_track.updated_at AS s_s_t_updated_at FROM sales_shipment_track GROUP BY parent_id ) AS s_s_t ON s_s_t.s_s_t_order_id = s_o.entity_id 
 
 GROUP BY s_o.entity_id
```

using `GROUP BY` without clauses in a subquery forces mysql to do this for every record in the shipment table for each query.with millions of records, this takes minutes instead of milliseconds. the same problems might exist for the other left joins, but in our case fixing the sales_shipment_track subquery was good enough. as the main query already uses a GROUP BY it's safe to remove the GROUP BY in the sales_shipment_track query.

**patch**
```
diff --git a/Model/ResourceModel/Document/Order/Entity.php b/Model/ResourceModel/Document/Order/Entity.php
--- a/Model/ResourceModel/Document/Order/Entity.php	(revision b150b5a48c0920e739e12016cddd5acd707f1740)
+++ b/Model/ResourceModel/Document/Order/Entity.php	(date 1641649055512)
@@ -27,7 +27,7 @@
         $mainEntitySelect = $this->getEntityByStoreIdsSelect($storeIds);
         $taxSelect = $this->appendPrefixToColumnsGroupBySelect("sales_order_tax", "s_o_t", "order_id");
         $paymentSelect = $this->appendPrefixToColumnsGroupBySelect("sales_order_payment", "s_o_p", "parent_id");
-        $shipmentTrackSelect = $this->appendPrefixToColumnsGroupBySelect("sales_shipment_track", "s_s_t", "parent_id");
+        $shipmentTrackSelect = $this->appendPrefixToColumnsSelect("sales_shipment_track", "s_s_t"); // group by is already done on the $mainEntitySelect
         $select = $this->adapter->select()
             ->from(
                 ['s_o' => new \Zend_Db_Expr("( ". $mainEntitySelect->__toString() . ' )')],
```

**side notes**
- using this mysql GROUP BY quirk to reduce a set of rows to pick a *random* one entry is considered bad practice as it can lead to non-deterministic behavior. mysql "randomly"-ish picks one of the matching records.
- if there is a 1:1 relation for tax & payments (which AFAIK is true in magento) their GROUP BY should be removed too
- is it really intentional that just one of the possibly (and realisticly) many sales_shipment_track rows per order is exported?

**Comment by dannegwork** [comment]
`is it really intentional that just one of the possibly (and realisticly) many sales_shipment_track rows per order is exported?`

Yes. Currently that information is not used for any logic. It is merely exported as a sign of "there is tracking".

**Comment by dannegwork** [comment]
`if there is a 1:1 relation for tax & payments (which AFAIK is true in magento) their GROUP BY should be removed too`

I agree. The group was added more as a preventive measure in case there is custom logic for payment & tax that is not 1:1 (as a fail-safe)

**Comment by dannegwork** [comment]
https://github.com/boxalino/data-integration-magento2/commit/d31ab5ffddc04ea8167ef435d422e2b7540e3f54

your fixes were applied. to be released in tag 1.2.3
